packages/affiliates/scripts/univ3.ts

```diff
@@ -0,0 +1,210 @@
+      descriptorProxyAddress: "0xd6852c52B9c97cBfb7e79B6ab4407AA20Ba31439",
+      nonfungibleTokenPositionManagerAddress: "0x2F9e608FD881861B8916257B76613Cb22EE0652c",
+      v3MigratorAddress: "0x03782388516e94FcD4c18666303601A12Aa729Ea"
+    }
+  ],
+  [
+    "ropsten",
+    {
+      v3CoreFactoryAddress: "0xDbe2c61E85D06eaA6E7916049f38B93288BA30f3",
+      weth9Address: "0xc778417E063141139Fce010982780140Aa0cD5Ab",
+      multicall2Address: "0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696",
+      proxyAdminAddress: "0xd3808aBF85aC69D2DBf53794DEa08e75222Ad9a1",
+      tickLensAddress: "0x9dF511178D1438065F7672379414F5C46D5B51b4",
+      quoterAddress: "0x2051F6Fb61077b5A2A2c17535d31A1F2C858994f",
+      swapRouter: "0x58f6b77148BE49BF7898472268ae8f26377d0AA6",
+      nonfungibleTokenPositionDescriptorAddress: "0xeb86f5BE368c3C5e562f7eA1470ACC431d30fB0C",
+      descriptorProxyAddress: "0xB79bDE60fc227217f4EE2102dC93fa1264E33DaB",
+      nonfungibleTokenPositionManagerAddress: "0x865F20efC14A5186bF985aD42c64f5e71C055376",
+      v3MigratorAddress: "0x1988F2e49A72C4D73961C7f4Bb896819d3d2F6a3"
+    }
+  ]
+]);
+const infura = process.env.infura;
+const network = "rinkeby";
+
+function getAddress(network: NetworkName, contractName: string) {
+  const addresses = networks.get(network);
+  assert(addresses[contractName], "invalid contract name: " + contractName);
+  return addresses[contractName];
+}
+
+async function getPoolState(params: { pool: Pool; provider: any }) {
+  const { pool, provider } = params;
+  assert(pool.address, "requires pool address");
+  const contract = new ethers.Contract(pool.address, UniswapV3Pool.abi, provider);
+  const slot0 = await contract.slot0();
+  const liquidity = await contract.liquidity();
+  return {
+    ...slot0,
+    liquidity: liquidity.toString()
+  };
+}
+async function processNftEvents(params: { provider: any; positions: ReturnType<typeof NftPositions> }) {
+  const { provider, positions } = params;
+  const nftHandler = NftEvents({ positions });
+  const contract = new ethers.Contract(
+    getAddress(network, "nonfungibleTokenPositionManagerAddress"),
+    NFTPositionManager.abi,
+    provider
+  );
+  const events = await contract.queryFilter({});
+  await Promise.map(events, nftHandler);
+}
+async function getNftPositionState(params: { provider: any; position: NftPosition }) {
+  const { provider, position } = params;
+  const contract = new ethers.Contract(
+    getAddress(network, "nonfungibleTokenPositionManagerAddress"),
+    NFTPositionManager.abi,
+    provider
+  );
+  return convertValuesToString(await contract.positions(position.tokenId));
+}
+async function processPoolEvents(params: {
+  provider: any;
+  pool: Pool;
+  positions: ReturnType<typeof Positions>;
+  pools: ReturnType<typeof Pools>;
+}) {
+  const { pools, pool, provider, positions } = params;
+  assert(pool.id, "requires pool id");
+  assert(pool.address, "requires pool address");
```

Comment:
These should never be empty right? If they can, can we check them at ingestion time rather than at processing time? Then we could make the type _require_ these fields and have the typechecker guarantee this for us.

Reply:
This goes back to the earlier point around making id required. Would be difficult particularly for positions which have a composite id based on owner, lowertick and uppertick, to have the user always pre calculate id.  Because typescript sees that these are optional, it forces us to assert they exist.  This is also a problem with partial updates, where events might include some data about the position, but not others.

IMO this would be  a really pain to fix, it requires making 2 types for every model, one that is a `Partial`, which is what essentially the types are now, and one that is a `Full` which has stronger guarantees around what fields exist.  Then you as the user must know which type to use based on the situation. Maybe you can see another way, happy to entertain suggestions.

Reply:
i just realized typescript has a `Partial<T>` utility. going to try this and see if i can make all fields required

Reply:
partial doesnt really solve things since everything ends up being a partial type. though i was able to apply this to the update function

Reply:
As discussed offline, I think there are some solutions to make _everything_ partial except specific required fields:
1. 
```ts
type PartialT = Partial<T> & {
  requiredField1: string;
  requiredField2: string;
};
```

2. 

```ts
type PartialExcept<T, K extends keyof T> = Partial<Omit<T, K>> & Required<Pick<T, K>>;
type PartialT = PartialExcept<T, 'requiredField1' | 'requiredField2'>;
```